Semantic analysis features

ishantheperson opened the pull request:
Let me know if you have any feedback on the general architecture before I get in too deep.

 - I added a "CreatesScope" to BlockStatement so that each group of statements keeps track of which variables are in scope there. This will also facilitate better autocompletion.

CalLavicka commented on server/src/ast.ts:
```diff
@@ -511,4 +525,164 @@ export type ConcreteType =
+export type Ordering = 
+  | "Less"
+  | "Equal"
+  | "Greater"; 
```
I think this should be a const enum rather than strings. Or, you could make it integer enums (so you could compare them with <, like order <= Ordering.Equal instead of comparing individually)

ishantheperson replied:
Good catch, for some reason I totally forgot enums existed

ishantheperson commented:
WordList might be useful as a backup, perhaps we can use it before we parse the document to an AST. But otherwise we are able to give better results using just the parsed results

ishantheperson commented:
Ok this is in a good state now
 - Added multi-file support using a config file
 - Added "go to definition" for local vars, can easily add functions
 - Added autocompletion for local vars, functions, and library functions (also need to add struct fields)
 - Added ability to see variable/struct field/function type information on hover

ishantheperson commented:
Multi file support should work now (with some caveats). See commit history
